src/runtime/cpu/kernels/kernels.cpp

```diff
@@ -78,3 +78,56 @@ float deepworks::CPUNLLLoss(Matrix predictions, ConstVector target) {
     float loss = -X_1d(0, slice).array().sum() / static_cast<float>(batch_size);
     return loss;
 }
+
+void deepworks::CPUBatchNorm1DForward(ConstMatrix input, Matrix output,
+                                      Matrix input_centered, Vector std,
+                                      Vector running_mean, Vector running_var,
+                                      bool isTraining, float eps, float alpha,
+                                      ConstVector gamma, ConstVector beta) {
+
+    if (isTraining) {
+        auto input_mean = input.colwise().mean();
+        input_centered = input.rowwise() - input_mean;
+
+        auto input_var = input_centered.cwiseAbs2().colwise().mean();
+        std = (input_var.array() + eps).cwiseSqrt();
+
+        running_mean = running_mean * alpha + input_mean * (1 - alpha);
+        running_var = running_var * alpha + input_var * (1 - alpha);
+
+        // FIXME: holy shit, if someone knows how to do it beautifully, tell me
+        output = ((input_centered.array().rowwise() / std.array())
+                 .array().rowwise() * gamma.array()).array().rowwise() + beta.array();
```

Comment:
the same, but less `.array().rowwise()`

```python
(input_centered.array().rowwise() * (gamma.array() / std.array())).array().rowwise() + beta.array();
```

Reply:
imho, it's not necessary to pass `Matrix input_centered, Vector std`. It is not difficult to count them in Backward, and the function's signature will become cleaner. But this is a moment for discussion.

also you can use (gamma.array() / std), if you set `auto` type for `input_centered` and `std`.

Reply:
type when use auto (BinaryOp) ![auto](https://user-images.githubusercontent.com/22177854/113171130-e101df00-924f-11eb-903b-9cb1fff834dc.jpg)

Reply:
if pass through the function ![pass_by](https://user-images.githubusercontent.com/22177854/113171350-0e4e8d00-9250-11eb-85d6-522198f0afe4.jpg)